x/clp/keeper/calculations.go

```diff
@@ -117,14 +141,31 @@ func CalculateWithdrawal(poolUnits sdk.Uint, nativeAssetBalance string,
 // slipAdjustment = (1 - ABS((R a - r A)/((2 r + R) (a + A))))
 // units = ((P (a R + A r))/(2 A R))*slidAdjustment
 
-func CalculatePoolUnits(oldPoolUnits, nativeAssetBalance, externalAssetBalance,
+func CalculatePoolUnits(symbol string, oldPoolUnits, nativeAssetBalance, externalAssetBalance,
 	nativeAssetAmount, externalAssetAmount sdk.Uint) (sdk.Uint, sdk.Uint, error) {
-	// refactor this to use ValidInputs
+	normalizationFactor := sdk.NewDec(1)
+	nf, ok := types.GetNormalizationMap()[symbol[1:]]
+	adjustExternalToken := true
+	if ok {
+		diffFactor := 18 - nf
+		if diffFactor < 0 {
+			diffFactor = nf - 18
+			adjustExternalToken = false
+		}
```

Comment:
How is this possible? Could we realistically have an asset that uses a larger precision than 1e18? (if I'm understanding this logic correctly).

Also, not a fan of just having numbers like these floating around... might be better to set as a constant (or better yet use the `Precision` constant in `github.com/cosmos/cosmos-sdk/client/types`) ... but guess that can be done another day.

Reply:
@intl-man this is not yet possible, but could be in the future if token with 18+ decimals is added, so I thought better to handle it right away. Agree about constants, there are many things including them that could be refactored

Reply:
@utx0 @intl-man btw, the token list in the repo assets has latest? Want to double check that we have all tokens with < 18 decimals in keys.go file, otherwise results can be incorrect